hs-bindgen/src-internal/HsBindgen/Frontend/Pass/Parse/Decl.hs

```diff
@@ -27,22 +29,26 @@ import HsBindgen.Language.C qualified as C
 -------------------------------------------------------------------------------}
 
 foldDecl :: HasCallStack => Fold ParseDecl [C.Decl Parse]
-foldDecl = foldWithHandler handleTypeException $ \curr ->
-    evalPredicate curr >>= \case
+foldDecl = foldWithHandler handleTypeException $ \curr -> do
+    info <- getDeclInfo curr
+    mNameKind <- dispatch curr $ return . C.toNameKindFromCXCursorKind
+    let mQualName = QualName <$> (C.isNamedDecl $ C.declId info) <*> mNameKind
+    evalPredicate (C.declLoc info) mQualName >>= \case
       Right () ->
+        -- TODO: This gets the CXCursorKind again.
         dispatchFold curr $ \case
           CXCursor_AlignedAttr        -> attribute
-          CXCursor_EnumDecl           -> enumDecl
-          CXCursor_FunctionDecl       -> functionDecl
+          CXCursor_EnumDecl           -> enumDecl info
```

Comment:
I tried to pass on the declaration info. Some functions are also used recursively, and so I didn't pass the top level `info` object, but let the function extract it's own `info` object. This may be a little slower than it could be, but is safer.

Reply:
Fixed, we now separate parse and skip actions.